src/stdlib/unflatten.rs

```diff
@@ -0,0 +1,329 @@
+use itertools::Itertools;
+
+use crate::compiler::prelude::*;
+
+static DEFAULT_SEPARATOR: &str = ".";
+
+fn unflatten(value: Value, separator: Value, recursive: Value) -> Resolved {
+    let separator = separator.try_bytes_utf8_lossy()?.into_owned();
+    let recursive = recursive.try_boolean()?;
+    let map = value.try_object()?;
+    Ok(do_unflatten(map.into(), &separator, recursive))
+}
+
+fn do_unflatten(value: Value, separator: &str, recursive: bool) -> Value {
+    match value {
+        Value::Object(map) => do_unflatten_entries(map, separator, recursive).into(),
+        // Note that objects inside arrays are not unflattened
+        _ => value,
+    }
+}
+
+fn do_unflatten_entries<I>(entries: I, separator: &str, recursive: bool) -> ObjectMap
+where
+    I: IntoIterator<Item = (KeyString, Value)>,
+{
+    let grouped = entries
+        .into_iter()
+        .map(|(key, value)| {
+            let (head, rest) = match key.split_once(separator) {
+                Some((key, rest)) => (key.to_string().into(), Some(rest.to_string())),
+                None => (key.clone(), None),
+            };
+            (head, rest, value)
+        })
+        .into_group_map_by(|(head, _, _)| head.clone());
+
+    grouped
+        .into_iter()
+        .map(|(key, mut values)| {
+            if values.len() == 1 {
+                match values.pop().unwrap() {
```

Comment:
```suggestion
            if values.len() == 1 {
                match values.pop().expect("exactly one element") {
```

We typically `expect` over `unwrap` to provide more context about why the panic should never be hit.

I think one potential way to refactor this that might avoid this unwrap/expect and avoid the early returns is something like:

```rust
match values {
  [value] => { ... what you currently have for handling if there is only one value ... }
  values => { ... handle multiple values ... }
}
```

Reply:
I tried to do that, but the problem is `values`  is a `Vec` and therefore, we should convert it to a slice in order to pattern match against it, using `values[..]`, but then, at the first branch we would have the first element borrowed and we would need it owned. The same with the second branch, we will have a borrowed slice of values, but not the owned values and we need owned values for the `do_unflatten_entries` function.

Does this makes sense?

Reply:
Mmm, yeah, that makes sense. I wish there was a better way to model this that doesn't require the `expect` but I'm not seeing it.